// gulp-wrap-commonjs/index.js
// Generated by CoffeeScript 1.10.0
(function() {
  'use strict';
  var PLUGIN_NAME, Replacer, _, applySourceMaps, defaultOptions, fs, isCoffeeScript, path, through;

  PLUGIN_NAME = 'gulp-wrap-commonjs';

  path = require('path');

  fs = require('fs');

  _ = require('lodash');

  through = require('through2');

  Replacer = require('regexp-sourcemaps');

  applySourceMaps = require('vinyl-sourcemaps-apply');

  defaultOptions = {
    autoRequire: false,
    moduleExports: false,
    relativePath: false,
    pathModifier: false,
    coffee: false
  };

  isCoffeeScript = function(filePath) {
    return filePath.slice(-7) === '.coffee';
  };

  module.exports = function(options) {
    var templateCoffee, templateJs;
    if (options == null) {
      options = {};
    }
    _.defaults(options, defaultOptions);
    templateJs = _.template(fs.readFileSync(__dirname + "/templates/js.tpl", {
      encoding: 'utf8'
    }));
    templateCoffee = _.template(fs.readFileSync(__dirname + "/templates/coffee.tpl", {
      encoding: 'utf8'
    }));
    return through.obj(function(file, enc, next) {
      var filePath, params, replacer, res, wrappedReplace;
      if (file.isBuffer()) {
        filePath = file.path;
        if (typeof options.pathModifier === "function") {
          filePath = options.pathModifier(file.path);
        }
        if (typeof options.relativePath === "string") {
          filePath = path.relative(path.join(process.cwd(), options.relativePath), filePath);
        }
        params = {
          filePath: filePath,
          autoRequire: options.autoRequire,
          moduleExports: options.moduleExports
        };
        if (options.coffee || isCoffeeScript(file.path)) {
          wrappedReplace = templateCoffee(params);
        } else {
          wrappedReplace = templateJs(params);
        }
        replacer = new Replacer(/^(.|\n)*$/, wrappedReplace, filePath + '.commonjs-wrap');
        res = replacer.replace(file.contents.toString('utf8'), file.relative);
        file.contents = new Buffer(res.code);
        if (file.sourceMap) {
          applySourceMaps(file, res.map);
        }
      }
      return next(null, file);
    });
  };

}).call(this);
